src/runtime/backend_loader.cpp

```diff
@@ -0,0 +1,178 @@
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#include "hipSYCL/runtime/backend_loader.hpp"
+#include "hipSYCL/common/debug.hpp"
+#include "hipSYCL/common/config.hpp"
+
+#include <cassert>
+#include <filesystem>
+#include <dlfcn.h>
+
+namespace {
+
+void close_plugin(void *handle) {
+  int err = dlclose(handle);
+
+  if (err != 0) {
+    HIPSYCL_DEBUG_ERROR << "backend_loader: dlclose() failed" << std::endl;
+  }
+}
+
+bool load_plugin(const std::string &filename, void *&handle_out,
+                 std::string &backend_name_out) {
+  void *handle = dlopen(filename.c_str(), RTLD_NOW);
+
+  if (!handle) {
+    HIPSYCL_DEBUG_ERROR << "backend_loader: Could not load backend plugin: "
+                        << filename << std::endl;
+    char *err = dlerror();
+    if (err) {
+      HIPSYCL_DEBUG_ERROR << err << std::endl;
+    }
+
+    return false;
+  } else {
+    void *symbol = dlsym(handle, "hipsycl_backend_plugin_get_name");
+
+    char *err = dlerror();
+    if (err) {
+      HIPSYCL_DEBUG_ERROR << "backend_loader: Could not retrieve backend name"
+                          << err << std::endl;
+    }
+    else {
+      auto get_name =
+          reinterpret_cast<decltype(&hipsycl_backend_plugin_get_name)>(symbol);
+
+      handle_out = handle;
+      backend_name_out = get_name();
+
+      return true;
+    }
+  }
+
+  // In case of failure, close handle to avoid resource leak
+  close_plugin(handle);
+  return false;
+}
+
+hipsycl::rt::backend *create_backend(void *plugin_handle) {
+  assert(plugin_handle);
+
+  void *symbol = dlsym(plugin_handle, "hipsycl_backend_plugin_create");
+  char *err = dlerror();
+  if (err) {
+    HIPSYCL_DEBUG_ERROR
+        << "backend_loader: Could not find symbol for backend creation" << err
+        << std::endl;
+    
+    return nullptr;
+  }
+  
+  auto create_backend_func =
+      reinterpret_cast<decltype(&hipsycl_backend_plugin_create)>(symbol);
+
+  return create_backend_func();
+}
+
+}
+
+namespace hipsycl {
+namespace rt {
+
+void backend_loader::query_backends() {
+  std::string install_prefix = HIPSYCL_INSTALL_PREFIX;
+
+  std::filesystem::path backend_lib_path =
+      std::filesystem::path{install_prefix} / "lib/hipSYCL";
+
+  std::string shared_lib_extension = ".so";
+  
+  for (const std::filesystem::directory_entry &entry :
+       std::filesystem::directory_iterator(backend_lib_path)) {
+
+    if (entry.is_regular_file()) {
+      auto p = entry.path();
+      if (p.extension().string() == shared_lib_extension) {
+        std::string backend_name;
+        void *handle;
+        if (load_plugin(p.string(), handle, backend_name)) {
+          HIPSYCL_DEBUG_INFO << "Successfully opened plugin: " << p
+                             << " for backend '" << backend_name << "'"
+                             << std::endl;
+          _handles.emplace_back(std::make_pair(backend_name, handle));
```

Comment:
with `emplace_back`, you should be able to omit the `std::make_pair` :)
`_handles.emplace_back(backend_name, handle);`

Reply:
hmm.. I don't really like the variadic template usage of `emplace_back` as it hides the relation of the arguments.